os/arch/xtensa/src/esp32/esp32_i2c.c

```diff
@@ -670,28 +701,47 @@ static void i2c_esp32_send_write_cmd(struct esp32_i2c_priv_s *priv, int addr_byt
 		*cmd++ = (struct i2c_esp32_cmd) {
 			.opcode = I2C_ESP32_OP_WRITE,
 			.num_bytes = to_send + addr_bytes,	//total len
-			.ack_en = true,
+			.ack_en = 1,
 		};
 
 		priv->data_count += to_send;
 		msg->length -= to_send;
 
 		*cmd = (struct i2c_esp32_cmd) {
 			.opcode = I2C_ESP32_OP_END,
-			.ack_en = false,
+			.ack_en = 0,
 		};
+		// end_detect will be disabled in isr handle, so enable it here.
+		I2C[priv->i2c_num]->int_ena.end_detect = 1;
 	}
-	
+
+	return to_send;
 }
 
-static void i2c_esp32_send_read_cmd(struct esp32_i2c_priv_s *priv, int addr_bytes, int cmd_start_index)
+static int i2c_esp32_send_read_cmd(struct esp32_i2c_priv_s *priv, int addr_bytes, int cmd_start_index)
 {
 	struct i2c_msg_s *msg = priv->transfer_msg;
 	volatile struct i2c_esp32_cmd *cmd = (void *)I2C_COMD0_REG(priv->i2c_num);
 	cmd += cmd_start_index;
 
-	uint32_t to_read = min(I2C_ESP32_BUFFER_SIZE, msg->length - 1);
-	volatile struct i2c_esp32_cmd *wait_cmd = NULL;
+	if (msg->length <= 0) {
+		priv->last_data_count = 0;
+		if (I2C_M_NORESTART != (msg->flags & I2C_M_NORESTART)) {
+			*cmd = (struct i2c_esp32_cmd) {
+				.opcode = I2C_ESP32_OP_STOP
+			};
+			return OK;
+		}
+
+		return ESP32_I2C_READ_ERROR;
+	}
+
+	int length = msg->length;
```

Comment:
Is it fine to declare a variable like this?

Reply:
yes, fixed. thanks.